src/fs/file.rs

```diff
@@ -40,12 +42,16 @@ pub struct File<'dir> {
     /// path (following a symlink).
     pub path: PathBuf,
 
+    /// The type of this file. None for directories `.` and `..`.
+    file_type: Option<FileType>,
```

Comment:
Did you check if it made a difference? Calling `file_type()` one time vs calling the various functions doesn’t poll the filesystem each time and doesn’t make a complex operation, [see there](https://github.com/rust-lang/rust/blob/543ab9964065dc2b4e47d7bff5e83c84519f265e/library/std/src/sys/unix/fs.rs#L270). Adding complexity isn’t worth it if this is optimized away by the compiler or if the different isn’t measurable.

Reply:
It's not for lazy. It's just that we can't construct it.
Normally `FileType` is from `DirEntry` when reading directories. As I said in comment, this field will be always `Some` for real `DirEntry` from `read_dir`, except for `.` and `..`, which are not returned by `read_dir` and are created as dummy instance with [these two functions](https://github.com/ogham/exa/pull/897/files#diff-c464eda590357178f19a588f108f7a09241fb6125aeaf8ea5acda2a501d06112L84-L105) .
 Since there is no real syscall, we cannot construct a `FileType`. I use `None` for these dummy `File` and make getter functions, like `is_directory`, correctly handle this case.

Reply:
But, couldn’t e.g. `self.file_type.map_or(false, |t| t.is_file())` be replaced by `self.metadata.file_type().map_or(false, |t| t.is_file())`, or I am missing something?

Sorry if it’s bothering you, but since it’s both a complicated part of the code and an important one, I have difficulty understanding everything that’s going on and I’d like not complicate it further if not strictly necessary. Your contribution is much appreciated. :slightly_smiling_face:

Reply:
`metadata` is expensive. This PR is just to elide them at all by use `file_type` as much possible, which is free from `DirEntry` without any syscall.
See also https://github.com/ogham/exa/pull/897#discussion_r659385287

Reply:
Ok, thank you for the explanation!